### PR TITLE
COO intersection: allow external hash + hash reuse in sparse_mask

### PR DESCRIPTION
**Pull request by nikitaved**
External hash implies more flexibility in the op coverage + performance improvement.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95165
* #94991
* __->__ #94596
* #94406

### CODE REVIEW

File: aten/src/ATen/native/sparse/SparseBinaryOpIntersectionCommon.h [hunk 1]
```diff
@@ -249,13 +267,22 @@ void _sparse_binary_op_intersection_kernel_impl(
...
     auto indices_nnz_stride = indices.stride(1);
 
-    auto hash = at::empty({probably_coalesced._nnz()},
-        indices.options().dtype(kLong));
+    auto hash = source_indices_hash_opt.has_value()
+      ? (*source_indices_hash_opt).contiguous()
+      : at::empty({probably_coalesced._nnz()}, indices.options().dtype(kLong));
+    const auto* RESTRICT hash_ptr = source_indices_hash_opt.has_value()
+      ? hash.data_ptr<int64_t>()
+      : nullptr;
```
Comment (nikitaved) [inline aten/src/ATen/native/sparse/SparseBinaryOpIntersectionCommon.h:285]: Bug: wrong place to put it.

File: aten/src/ATen/native/sparse/SparseStubs.h [hunk 2]
```diff
@@ -2,17 +2,20 @@
...
 #include <c10/util/ArrayRef.h>
+#include <c10/util/Optional.h>
 
 namespace at {
 
 class Tensor;
 
 namespace native {
 
+using OptTensor = c10::optional<Tensor>;
```
Comment (cpuhrsch) [inline aten/src/ATen/native/sparse/SparseStubs.h:13]: Nit: Adding this to the header makes it too widely available.

**Comment by nikitaved** [comment]
@cpuhrsch , yes, it allows to implement intersections along a specific dimension, and so is broadcasting all with the same kernel.

**Review by cpuhrsch — APPROVED** [review]
Can we remove `using OptTensor` from the headers? Otherwise this is good to go.

File: aten/src/ATen/native/sparse/SparseBinaryOpIntersectionCommon.h [hunk 3]
```diff
@@ -146,10 +148,23 @@ void _sparse_binary_op_intersection_kernel_impl(
...
-  const Tensor x = distributive_with_sum ? x_ : x_.coalesce();
-  const Tensor y = distributive_with_sum ? y_ : y_.coalesce();
+  const auto coalesce_if_not_distributive = [distributive_with_sum](const Tensor& t, const OptTensor& t_hash_opt) -> auto {
+    // No need to coalesce in such a case.
+    if (distributive_with_sum) {
+      return std::make_tuple(t, t_hash_opt);
+    } else {
+      // Otherwise coalesce and force hash recompute.
+      return std::make_tuple(t.coalesce(), static_cast<OptTensor>(c10::nullopt));
+    }
```
Comment (nikitaved) [inline aten/src/ATen/native/sparse/SparseBinaryOpIntersectionCommon.h:161]: TODO: no need to recompute, just sufficient to wrap into a sparse tensor and coalesce.

**Pull request closed without merging**